### PR TITLE
[runtime] MonoError-ize mono_object_isinst{,_mbyref}

### PR DESCRIPTION
**Pull request by lambdageek**
Also mark them external only.  Runtime should use mono_object_isinst_mbyref_checked and mono_object_isinst_checked

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -250,6 +253,7 @@ ves_icall_System_Array_SetValueImpl (MonoArray *arr, MonoObject *value, guint32
 
 #define INVALID_CAST G_STMT_START{ \
 		mono_get_runtime_callbacks ()->set_cast_details (vc, ec); \
+		mono_error_cleanup (&error);				\
```
Comment (kumpera) [inline mono/metadata/icall.c]: We should raise the underlying exception and not hide it under an InvalidCastException

File: mono/metadata/icall.c [hunk 2]
```diff
@@ -289,19 +293,23 @@ ves_icall_System_Array_SetValueImpl (MonoArray *arr, MonoObject *value, guint32
...
-	if (mono_object_isinst (value, ec)) {
+	if (mono_object_isinst_checked (value, ec, &error)) {
 		if (ec->has_references)
 			mono_value_copy (ea, (char*)value + sizeof (MonoObject), ec);
 		else
 			mono_gc_memmove_atomic (ea, (char *)value + sizeof (MonoObject), esize);
 		return;
 	}
+	if (!is_ok (&error)) {
+		mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/metadata/icall.c]: This is cleaning up the same error twice.

We should kill that INVALID_CAST macro and replace with something that takes no implicit state.

File: mono/metadata/icall.c [hunk 3]
```diff
@@ -3063,7 +3076,8 @@ ves_icall_InternalInvoke (MonoReflectionMethod *method, MonoObject *this_arg, Mo
 		}
 
 		if (this_arg) {
-			if (!mono_object_isinst (this_arg, m->klass)) {
+			if (!mono_object_isinst_checked (this_arg, m->klass, &error)) {
+				mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/metadata/icall.c]: We shouldn't swallow this error.

File: mono/mini/exceptions-x86.c [hunk 4]
```diff
@@ -473,13 +474,14 @@ mono_x86_throw_exception (mgreg_t *regs, MonoObject *exc,
...
 
-	if (mono_object_isinst (exc, mono_defaults.exception_class)) {
+	if (mono_object_isinst_checked (exc, mono_defaults.exception_class, &error)) {
 		MonoException *mono_ex = (MonoException*)exc;
 		if (!rethrow) {
 			mono_ex->stack_trace = NULL;
 			mono_ex->trace_ips = NULL;
 		}
 	}
+	mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/exceptions-x86.c]: This should be an assert, we shouldn't fail type loading for allocated objects.

File: mono/metadata/object.c [hunk 5]
```diff
@@ -5970,18 +5970,45 @@ mono_object_unbox (MonoObject *obj)
...
+ * Returns: @obj if @obj is derived from @klass or NULL otherwise.
  */
 MonoObject *
 mono_object_isinst (MonoObject *obj, MonoClass *klass)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
+	MonoError error;
+	MonoObject *result = mono_object_isinst_checked (obj, klass, &error);
+	mono_error_cleanup (&error); /* FIXME better API that doesn't swallow the error */
```
Comment (kumpera) [inline mono/metadata/object.c]: If the method now is banned from the API, the comment can go.

File: mono/mini/jit-icalls.c [hunk 6]
```diff
@@ -102,11 +102,13 @@ mono_ldvirtfn_gshared (MonoObject *obj, MonoMethod *method)
...
 mono_helper_stelem_ref_check (MonoArray *array, MonoObject *val)
 {
+	MonoError error;
 	if (!array) {
 		mono_set_pending_exception (mono_get_exception_null_reference ());
 		return;
 	}
-	if (val && !mono_object_isinst (val, array->obj.vtable->klass->element_class)) {
+	if (val && !mono_object_isinst_checked (val, array->obj.vtable->klass->element_class, &error)) {
+		mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/jit-icalls.c]: Don't swallow the error

File: mono/mini/jit-icalls.c [hunk 7]
```diff
@@ -1192,8 +1195,9 @@ mono_object_castclass_unbox (MonoObject *obj, MonoClass *klass)
 	oklass = obj->vtable->klass;
 	if ((klass->enumtype && oklass == klass->element_class) || (oklass->enumtype && klass == oklass->element_class))
 		return obj;
-	if (mono_object_isinst (obj, klass))
+	if (mono_object_isinst_checked (obj, klass, &error))
 		return obj;
+	mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/jit-icalls.c]: Same thing, assert of throw, but don't swallow.

File: mono/mini/jit-icalls.c [hunk 8]
```diff
@@ -1226,10 +1231,11 @@ mono_object_castclass_with_cache (MonoObject *obj, MonoClass *klass, gpointer *c
 	if (cached_vtable == obj_vtable)
 		return obj;
 
-	if (mono_object_isinst (obj, klass)) {
+	if (mono_object_isinst_checked (obj, klass, &error)) {
 		*cache = obj_vtable;
 		return obj;
 	}
+	mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/jit-icalls.c]: Don't swallow the error.

File: mono/mini/jit-icalls.c [hunk 9]
```diff
@@ -1257,10 +1264,11 @@ mono_object_isinst_with_cache (MonoObject *obj, MonoClass *klass, gpointer *cach
 		return (cached_vtable & 0x1) ? NULL : obj;
 	}
 
-	if (mono_object_isinst (obj, klass)) {
+	if (mono_object_isinst_checked (obj, klass, &error)) {
 		*cache = (gpointer)obj_vtable;
 		return obj;
 	} else {
+		mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/jit-icalls.c]: We shouldn't swallow this error and silently fail the type check.

File: mono/mini/mini-exceptions.c [hunk 10]
```diff
@@ -1373,10 +1374,11 @@ mono_handle_exception_internal_first_pass (MonoContext *ctx, MonoObject *obj, gi
 	mono_ex = (MonoException*)obj;
 	initial_trace_ips = mono_ex->trace_ips;
 
-	if (mono_object_isinst (obj, mono_defaults.exception_class)) {
+	if (mono_object_isinst_checked (obj, mono_defaults.exception_class, &error)) {
 		mono_ex = (MonoException*)obj;
 		initial_trace_ips = mono_ex->trace_ips;
 	} else {
+		mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/mini-exceptions.c]: This should be an assert.

File: mono/mini/mini-exceptions.c [hunk 11]
```diff
@@ -1605,7 +1608,8 @@ mono_handle_exception_internal (MonoContext *ctx, MonoObject *obj, gboolean resu
 		obj = (MonoObject *)mono_get_exception_null_reference ();
 	}
 
-	if (!mono_object_isinst (obj, mono_defaults.exception_class)) {
+	if (!mono_object_isinst_checked (obj, mono_defaults.exception_class, &error)) {
+		mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/mini-exceptions.c]: Make it an assert.

File: mono/mini/mini-exceptions.c [hunk 12]
```diff
@@ -3051,10 +3059,13 @@ mono_llvm_match_exception (MonoJitInfo *jinfo, guint32 region_start, guint32 reg
 		}
 
 		// FIXME: Handle edge cases handled in get_exception_catch_class
-		if (ei->flags == MONO_EXCEPTION_CLAUSE_NONE && mono_object_isinst (exc, catch_class)) {
+		if (ei->flags == MONO_EXCEPTION_CLAUSE_NONE && mono_object_isinst_checked (exc, catch_class, &error)) {
 			index = ei->clause_index;
 			break;
-		} else if (ei->flags == MONO_EXCEPTION_CLAUSE_FILTER) {
+		} else
+			mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/mini-exceptions.c]: Assert instead.

File: mono/mini/mini-exceptions.c [hunk 13]
```diff
@@ -2882,7 +2889,8 @@ throw_exception (MonoObject *ex, gboolean rethrow)
 	MonoJitTlsData *jit_tls = mono_get_jit_tls ();
 	MonoException *mono_ex;
 
-	if (!mono_object_isinst (ex, mono_defaults.exception_class)) {
+	if (!mono_object_isinst_checked (ex, mono_defaults.exception_class, &error)) {
+		mono_error_cleanup (&error);
```
Comment (kumpera) [inline mono/mini/mini-exceptions.c]: Assert instead.

**Comment by kumpera** [comment]
My understanding is that mono_object_isinst_checked and friends are functions that return 3 possible values: true, false and false + error.

We should not systematically ignore the last case.

**Comment by lambdageek** [comment]
Thanks @kumpera.  I updated the PR to cover the false+error cases.

Also I missed all the non-OSX platforms, so I also updated the PR to fixup the other `exceptions-{arch}.c`  files.

**Comment by lambdageek** [comment]
@monojenkins merge

**Pull request merged**